Inline notifications (#3259)
grommet/hpe-design-system: aries-site/src/pages/templates/inline-notifications.mdx

import { Example } from '../../layouts';
import { InlineNotificationExample } from '../../examples';
import { Checkmark } from 'grommet-icons';

Inline notifications are messages that appear near the content to which they
are related. They may be task or system generated, communicating information
such as status updates, alerts to success or failure of an action, or call
attention to a new feature in which a user may be interested.

<Example
  code=""
  github=""
  showResponsiveControls={['fullScreen']}
  screenContainer
  height="650px"
  width="100%"
>
  <InlineNotificationExample />
</Example>

## When to use

Consider using an inline notification when presenting:

- **Feedback** &ndash; Providing feedback to users about their actions. For
  example, if a user submits a form requiring server-side validation, an
  inline notification should be used to display and summarize form-level errors.
- **Significant status updates** &ndash; Delivering significant status updates

>   about a task. For example, if a user initiated an ansynchronous process such

Reviewer: Done.

  as a device update, a notification might provide the status update upon.
  "Success - all devices have updated to the latest version."

> - **Promotions** &ndash; Calling a user's attention to such as new features, tips,

Reviewer: Should we flag that this should be used sparingly? Or responsibly? Something to indicate it shouldn't be reached for all the time (I think people know that but just in case)

Reviewer: Done.

  best practices, or other announcements.

## When to consider something else

Inline notifications are quite flexible, however consider using the alternatives
for the following scenarios:

- **Low severity information** &ndash; If communicating low severity information
  which does not need to persist or require the user's attention, consider using

>   a [toast notification](/templates/toast-notifications).

Author: @taysea, on the /components/notification page, can we add a cross link to the inline notifications page?

Also, can we add an issue for "Notifications - Add decision tree to notification page for when to use each type of notification'?

- **High attention information** &ndash; Consider using [global banner notification](/templates/global-banner-notifications)
  when communicating information which might result in loss of functionality or
  requires the user's attention and/or action. Examples might include a system
  outage, critical error, or expiring subscription or license. [Global banner notifications](/templates/global-banner-notifications)
  are persistent and require user action to acknowledge or dismiss.
- **Immediate feedback** &ndash; When providing immediate feedback to a user's
  action, a variety of tools are likely more appropriate.

  - When a user is awaiting a response after clicking an action button, use the

>     [busy button]() which provides in-progress and success animations. If the

Reviewer: @halocline do you want me to add a busy button example on the button page to link to?

>     [busy button]() which provides in-progress and success animations. If the

Reviewer 2: FYI I created an issue to add a button busy example to the site https://github.com/grommet/hpe-design-system/issues/3265

    response time exceeds expectations or the API call fails, then an inline
    notification is appropriate for providing status updates.
  - When validating form input client-side, use [field-level form validation]()
    when possible. If validation results in a form-level error, using an inline
    notification is appropriate.

## Anatomy

An inline notification consists of:

| Label | Region                       | Purpose                                                      |                  Required                   | Notes                                                                                                                  |
| ----- | ---------------------------- | ------------------------------------------------------------ | :-----------------------------------------: | ---------------------------------------------------------------------------------------------------------------------- |
| **1** | **Icon or status indicator** | Indicates the type of notification                           | <Checkmark a11yTitle="true" size="small" /> | <span>Normal, Warning, Critical, and Unknown [status indicators](/templates/status-indicator) can be displayed.</span> |

> | **2** | **Title**                    | Short, descriptive label identifying the notification        |                                             | Titles should only be used for inline notifications presenting page level summaries.                                   |

Reviewer: In v5 of the theme, title uses weight 500

| **3** | **Message**                  | Briefly describes the action, status, or additional context  | <Checkmark a11yTitle="true" size="small" /> |                                                                                                                        |

> | **4** | **Anchor**                   | Used to direct users to additional detail or related content |                                             |                                                                                                                        |

Author: Addressing in suggestion below

> | **5** | **Close button**             | Dismisses the notification                                   |                                             |                                                                                                                        |

Reviewer: Should we add a note here that Close button should not be used in "Critical" notifications?

Reviewer: Done.

## Placement

Inline notifications should be placed near the content to which they are
related.

- If the notification is related to a specific element or section of content,

>   it should be placed near that element or section of content.

Author: Good clarification. Will do.

>   it should be placed near that element or section of content.

Reviewer: Done.

- If the notification is related to the entire page, it should be placed near

>   the top of the page.

Author: Agree. More specificity here would be valuable.

Will do.

>   the top of the page.

Reviewer: Is this a good candidate for a "Do" "Dont" example?

>   the top of the page.

Reviewer: For now, I think the wording change is sufficient. We can always add a do don't later but don't think it's necessary for mvp.

>   the top of the page.

Reviewer: Changed wording.

  notification. The links should immediately follow the message and should be
  limited to no more than two links.

## Content guidelines

Notifications reassure users that they are moving towards their goal. They should
be relevant, timely, and informative providing assurance that their intended
actions have been completed successfully. Or, if there is an issue, they should
provide clear direction on how to resolve the issue.

- **Relevant** &ndash; Notifications should be relevant to the user's current
  task.
- **Timely** &ndash; Deliver notifications in a timely manner and ensure they
  are kept up to date with current status.
- **Informative** &ndash; Messages shoud provide clear context and direction on
  how to resolve any issues. Follow the [voice and tone guidelines](/guidelines/voice-and-tone) for writing
  crisp and effective content.

## Behavior

- Never sticky and should scroll when there is other content.
- The width and resizing behavior of an inline notification should scale alongside
  related content. For example, when placed below the [PageHeader](/components/pageheader),
  the inline notification should have the same width and responsive behavior as the

>   PageHeader.

Reviewer 3: I think we want to showcase this through one of the examples/use cases

>   PageHeader.

Reviewer: Done. This is captured in the "Additional detail" example.

- Inline Notifications are not clickable. Instead use anchors to direct users to
  take further action or to view the issue in full detail.
- Do not place buttons within an inline notification.

## Do's and dont's

## Use cases

### Status updates

> Inline notifications may be used to display okay, warning, and critical information.

Reviewer 3: ```suggestion
Inline notifications may be used to display normal, warning, and critical information.
```

We're currently using the terminology 'normal' instead of 'okay' within our status indicators guidance.

Author: 👍

Reviewer: Done.

> For example, an okay status might communicate a process is progressing

Reviewer 3: ```suggestion
For example, a normal status might communicate a process is progressing
```

Author: 👍

Reviewer: Done.

> without interuption, a warning status might communicate a resource is approaching

Reviewer: ```suggestion
without interruption, a warning status might communicate a resource is approaching
```

  best practices, or other announcements. Use these sparingly.

## When to consider something else

Inline notifications are quite flexible, however consider using the alternatives
for the following scenarios:

- **Low severity information** &ndash; If communicating low severity information
  which does not need to persist or require the user's attention, consider using
  a [toast notification](/templates/toast-notifications).
- **High attention information** &ndash; Consider using [global banner notification](/templates/global-banner-notifications)
  when communicating information which might result in loss of functionality or
  requires the user's attention and/or action. Examples might include a system
  outage, critical error, or expiring subscription or license. [Global banner notifications](/templates/global-banner-notifications)
  are persistent and require user action to acknowledge or dismiss.
- **Immediate feedback** &ndash; When providing immediate feedback to a user's
  action, a variety of tools are likely more appropriate.

  - When a user is awaiting a response after clicking an action button, use the

>     [busy button]() which provides in-progress and success animations. If the

Reviewer: ```suggestion
    [busy button](/components/button#busy-button) which provides in-progress and success animations. If the
```
Pending merge of https://github.com/grommet/hpe-design-system/pull/3344

| ----- | ---------------------------- | ------------------------------------------------------------ | :-----------------------------------------: | ---------------------------------------------------------------------------------------------------------------------- |
| **1** | **Icon or status indicator** | Indicates the type of notification                           | <Checkmark a11yTitle="true" size="small" /> | <span>Normal, Warning, Critical, and Unknown [status indicators](/templates/status-indicator) can be displayed.</span> |
| **2** | **Title**                    | Short, descriptive label identifying the notification        |                                             | Titles should only be used for inline notifications presenting page level summaries.                                   |
| **3** | **Message**                  | Briefly describes the action, status, or additional context  | <Checkmark a11yTitle="true" size="small" /> |                                                                                                                        |

> | **4** | **Anchor**                   | Used to direct users to additional detail or related content |                                             |                                                                                                                        |

Author: ```suggestion
| **4** | **Link**                   | Used to direct users to additional detail or related content |                                             |                                                                                                                        |
```
